KEP: FailureRecoveryPolicy for gracefully handling "stuck" terminating Pods (#7311)
kubernetes-sigs/kueue: keps/6757-failure-recovery/README.md

as my framework expects exactly one pod per worker index.

Some nodes in my cluster often go offline for long periods of time, during which my job is unable to make progress -
replacement pods will not be scheduled as long as the node remains offline. I'd like for my jobs to be automatically unblocked
as soon as possible, so I can make best use of the available resources (since the training process is synchronized,
there is no risk of data corruption in my case).

### Risks and Mitigations

#### Replacing Pods Which Are Still Running

Starting a replacement pod without a guarantee that the original one was properly terminated carries an inherent risk
and might result in inconsistencies, as discussed above.
If this risk is not communicated to the user properly and the feature is not sufficiently explicit in it's behavior,
the users might mistake it for a general issue within Kueue/the broader Kubernetes ecosystem.

As this risk stems from the behavior of the `kubelet` and control plane themself, it cannot be fully mitigated without major changes in how Kubernetes operates.

> Instead, it should be adequately documented to prevent users without a compatible use-case from using it.

Reviewer: Add that as a mitigation we make this behavior opt-in on two levels: admin needs to enable in the configuration, and users need to create Pods with labels matching the configured label selector by the admin.

The controller has to **ignore** updates to pods that:
1. Are not terminating.
    * `pod.DeletionTimestamp == nil || pod.DeletionGracePeriodSeconds == nil`
1. Are in a terminal phase.
    * `pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodPending`
1. Are not managed by Kueue.
    * Neither the `kueue.x-k8s.io/managed` (for pod integration) nor `kueue.x-k8s.io/podset` label is set.
1. Don't match the label selector defined in the configured `FailureRecoveryPolicy`.
1. Are **not** scheduled on a node tainted with `node.kubernetes.io/unreachable`.
    * This explicitly ignores pods assigned to nodes that still have a running kubelet.
    For example, nodes with the `node.kubernetes.io/not-ready` taint experiencing resource pressure
    that makes pod termination take longer.

For relevant (not ignored) terminating pods, the controller schedules another reconciliation
to happen after the remaining grace period elapses:

```go
func (r *ZombiePodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
  // ...

	now := r.clock.Now()
	gracefulTerminationPeriod := time.Duration(*pod.DeletionGracePeriodSeconds) * time.Second

>   forcefulTerminationGracePeriod := time.Duration(*terminatePodConfig.ForcefulTerminationGracePeriod) * time.Second

Reviewer 2: `*tPC.FTGP` already is a `time.Duration`.

Also, nit: fix indent.

i.e. pods that are stuck in the `Pending`/`Running` state due to a node-level failure.

## Motivation

Currently, a malfunction of the `kubelet` (or a more general node failure) results in the pods running on that node to become stuck.
When the `kubelet` fails to send its regular heartbeat to the control plane within `node-monitor-grace-period`, the node is deemed unhealthy and is assigned the `node.kubernetes.io/not-ready` taint. By default, Kubernetes [automatically adds a 5-minute toleration](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions) for to pods running in the affected node (this can be explicitly set by a different value by the user/controller).
If the heartbeat is resumed within this toleration period, the control plane removes the taints and the pods continue running.
On the other hand, if it is not resumed, then the pods are evicted from the node and marked for termination (i.e. their `deletionTimestamp` is set in `etcd`).

Without a functioning `kubelet`, the pods have no way of progressing beyond that point on their own.

> Such pods, colloquially called "zombie pods", remain terminating until the node is healed or they are manually removed.

Author: Good questions. By pointing this out, I noticed that the `unreachable` taint should be mentioned in the KEP, not `not-ready`.

The node assignment is not removed when a pod is being deleted, so the pods are still "scheduled" i.e. do not fall under criterion # 2. You got it right - the `unreachable` taint is different than the `out-of-service` taint, so criterion # 3 does not apply.

As for your follow-up - yes, I think we'd like to do that. Initially I wanted to base it off solely off pods, but now I noticed another benefit (apart from restricting scope) - it will explicitly ignore pods assigned to a node that is `not-ready` (for example due to pressure on some resource) and terminating for a long time because of that. If the node is `not-ready`, the kubelet is still running and communicating with the control plane.

### Risks and Mitigations

#### Replacing Pods Which Are Still Running

Starting a replacement pod without a guarantee that the original one was properly terminated carries an inherent risk
and might result in inconsistencies, as discussed above.
If this risk is not communicated to the user properly and the feature is not sufficiently explicit in its behavior,
the users might mistake it for a general issue within Kueue/the broader Kubernetes ecosystem.

As this risk stems from the behavior of the `kubelet` and control plane themself, it cannot be fully mitigated without major changes in how Kubernetes operates.
Instead, it should be adequately documented to prevent users without a compatible use-case from using it.
Moreover, enabling this feature should require opt-in both from:
1. The Administrator - by defining the recovery rules and selectors in the Kueue config.
2. The User - by opting specific workloads into the failure recovery mechanism.

Controlling the covered workloads, instead of applying the recovery globally, will guarantee that only pods that were deemed "safe to forcefully terminate" are affected.

## Design Details

### Configuration API

> The `Configuration` struct is extended to add `FailureRecoveryPolicies`:

Reviewer 2: ```suggestion
The `Configuration` struct is extended to add `FailureRecoveryPolicy`:
```

The proposal is to **not** set a default value for the forceful termination grace period.
This puts configuration over convention, as it will make the user consciously think about the
grace period the need for their specific use case, without risking a faulty default.

Since enabling the feature is inherently risky, this serves as an additional explicit
acknowledgement of this risk from the user's end.

Alternatively, a default value can be inferred from the `terminationGracePeriodSeconds`.
For example, it could be a multiple of that value. Nevertheless, this still runs the risk of
setting a very small value or even 0, depending on the user's configuration.

### Controller

Setting the strategy type to `PodTermination` will enable a controller, which manages
the failed nodes.

The controller has to ignore any updates to pods that:
1. Are not terminating.
    * `pod.DeletionTimestamp == nil || pod.DeletionGracePeriodSeconds == nil`
1. Are in a terminal phase.
    * `pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodPending`
1. Are not managed by Kueue.
    * The `kueue.x-k8s.io/managed` label is set for pods with pod integration enabled.

>     * The presence of the `kueue.x-k8s.io/podset` label could be used in other cases.

Reviewer 2: ```suggestion
1. Are not managed by Kueue.
    * That means, have neither of the labels `kueue.x-k8s.io/managed` and `kueue.x-k8s.io/podset` set.
```

This would be clearer to me - especially given your convention from the bullets above.
(Or, if it sounds less clear to you, please at least tell me if I got it right :)

1. Match the label selector defined in the configured `FailureRecoveryPolicy`.

For terminating pods matching the criteria, the controller schedules another reconciliation
to happen after the remaining grace period elapses:

```go
func (r *ZombiePodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
  // ...

	now := r.clock.Now()
	gracefulTerminationPeriod := time.Duration(*pod.DeletionGracePeriodSeconds) * time.Second

> 	totalGracePeriod := gracefulTerminationPeriod + zombiePodTerminationGracePeriod

Reviewer 2: Is `zombiePodTGP` same as `forcefulTGP` declared above?

	if now.Before(pod.DeletionTimestamp.Add(totalGracePeriod)) {
		gracePeriodLeft := pod.DeletionTimestamp.Add(totalGracePeriod).Sub(now)
		return ctrl.Result{RequeueAfter: gracePeriodLeft}, nil
	}

  // ...
}
```

In that scheduled reconciliation, unless the node recovered or the pod was deleted,
the pod will be deemed "zombie" and deleted:

```go
func (r *ZombiePodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
  // ...

	pod.Status.Phase = corev1.PodFailed

> 	if err := r.client.Status().Update(ctx, pod); err != nil {

Reviewer: sgtm, we should aim for minimal support in the first iteration, but flexible API which will allow to introduce knobs like "delete" vs mark "Failed". I think it is reasonable to assume that the default in the first iteration is just "mark as Failed".

## Alternatives

### Do Nothing

Since this feature spawned a lengthy discussion about its riskiness and there are other controllers in the ecosystem which deal with
node problem remediation (e.g. [medik8s](https://github.com/medik8s/self-node-remediation)), an alternative would be to do nothing,
wait for the upstream conversations to be resolved and propose an alternative solution (external to Kueue) to the affected users.

#### Advantages

1. Requires no implementation effort.

#### Drawbacks

1. Forces users to run another system alongside Kueue, making deploying Kueue more complex those cases.

### Hide The Controller Behind A Feature Gate

Instead of introducing changes to the API, a feature gate could be used to get some initial feedback about the
feature and its risks.

#### Advantages

1. Easy to implement, less moving parts.
1. Allows to gather feedback quickly.

#### Drawbacks

1. It won't allow to implement the pod selectors and configurations for the failure recovery,
limiting the implementation to defaults:
  1. `podLabelSelector` - everything.

>   1. `forcefulTerminationGracePeriod` - a constant or a multiple of `terminationGracePeriodSeconds`.

Reviewer 2: This misrenders, please fix indents.

>   1. `forcefulTerminationGracePeriod` - a constant or a multiple of `terminationGracePeriodSeconds`.

Author: It would be `N * terminationGracePeriodSeconds`, where N is a constant. It's one of the ideas discussed in https://github.com/kubernetes/kubernetes/issues/134038, but I did not expand on it here.

Do you think it warrants more explanation?

>   1. `forcefulTerminationGracePeriod` - a constant or a multiple of `terminationGracePeriodSeconds`.

Reviewer 2: Oh, never mind, I imagined some non-existing problem.

Core Kubernetes already contains logic for [automatic garbage collection](https://github.com/kubernetes/kubernetes/blob/4870d987d0a4aac2d9223d4c0b9f22858c0d1590/pkg/controller/podgc/gc_controller.go) of `Pod`s running on a `Node` with the `node.kubernetes.io/out-of-service` taint.
It updates the `Pod`s status and deletes it from `etcd`. The recovery controller could use this fact to terminate stuck pods by automatically adding this taint to nodes which are not ready for some configurable time.

Given the API proposal, this approach could potentially be implemented as an alternative
 recovery strategy in the future.

 ```go
type OutOfServiceNodeTaintingConfig struct {
  NotReadyNodeTimeout *time.Duration `json:"notReadyNodeTimeout"`
}

type FailureRecoveryAction string

const (
  PodTermination FailureRecoveryAction = "PodTermination"
  TaintingNodeOutOfService FailureRecoveryAction = "TaintingNodeOutOfService"
)
 ```

#### Advantages

1. It uses an existing Kubernetes mechanism.
2. Concerns are tidily separated - the recovery controller simply says the node is out of service, the garbage collector in the control plane can decide how and when to clean it up.

> #### Drawbacks

Reviewer: nit: I would suggest not to use the "Drawbacks" subsection, because there is already such section referring to the selected approach. Just to avoid a reader getting tricked into counting the nesting levels. For the "Alterneatives" section we usually just have the "reasons to defer" paragraph (take a look at the TAS KEP).

}

type FailureRecoveryPolicy struct {
  // Rules specifies the rules to be enabled for failure recovery.
  // +kubebuilder:validation:MinItems=1
  Rules []FailureRecoveryRule `json:"rules"`
}

// +kubebuilder:validation:ExactlyOneOf=["terminatePod"]
type FailureRecoveryRule struct {
  // Exactly one of the fields below must be specified.

  // TerminatePod enables and contains configuration for the `TerminatePod` strategy.
  // This strategy recovers stuck pods by forcefully terminating them after a configured
  // grace period elapses.
  // +optional
  TerminatePod *TerminatePodConfig `json:"terminatePod,omitempty"`
}

type TerminatePodConfig struct {
  // PodLabelSelector specifies the scope of resources covered by `TerminatePod` failure recovery -
  // resources not matching the selector are ignored by the controller.
  // If a pod matches multiple configurations, the strictest one (with the shortest grace period)
  // will be applied.
  PodLabelSelector *metav1.LabelSelector `json:"podLabelSelector,omitempty"`

  // ForcefulTerminationGracePeriod is the duration between when the pod's `deletionGracePeriodSeconds`

>   // elapses and when the pod should be forcefully deleted.

Reviewer: This is more of the description about evaluating the order of rules, so the comment belongs to the "Rules" field inside FailureRecoveryPolicy.

However, I'm not certain this approach is the best. Alternatives I see:
1. we choose whatever is the first matching rule
2. we choose the least disruptive action first (say PodTermination over taining Node)
3. we choose the first in time action (as proposed currently)

I would suggest to avoid committing and for now only support a single rule.

  // Represented using metav1.Duration (e.g. "10m", "1h30m").
  // +required
  ForcefulTerminationGracePeriod metav1.Duration `json:"forcefulTerminationGracePeriod"`
}
```

This API requires opt-in both from the administrators and user's end, as the workloads would
have to be labeled in order for them to be covered with failure recovery.

An example failure recovery configuration:

```yaml
failureRecoveryPolicy:
  rules:
  - terminatePod:
      podLabelSelector:
        matchExpressions:
        - key: safe-to-fail
          operator: In

>           values: [ "true", "True", "yes" ]

Reviewer: nit, let's just use one value
```suggestion
          values: [ "true" ]
```

      forcefulTerminationGracePeriod: 5m
```

### Default Grace Period

The proposal is to **not** set a default value for the forceful termination grace period.
This puts configuration over convention, as it will make the user consciously think about the
grace period the need for their specific use case, without risking a faulty default.

Since enabling the feature is inherently risky, this serves as an additional explicit
acknowledgement of this risk from the user's end.

Alternatively, a default value can be inferred from the `terminationGracePeriodSeconds`.
For example, it could be a multiple of that value. Nevertheless, this still runs the risk of
setting a very small value or even 0, depending on the user's configuration.

> ### Controller

Reviewer: ```suggestion
### Implementation overview
```

}
```

### Test Plan

[ ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit Tests

The proposal will be covered with unit tests for:
1. Configuration parsing.
1. Controller behavior:
    1. Whether it ignores irrelevant pods (not terminating, not managed by Kueue, already failed/succeeded, not labeled).
    1. Whether it correctly schedules a reconciliation for when the grace period elapses.
    1. Whether it updates the pod's phase to `Failed` after the grace period elapses.

#### Integration Tests

The proposal will be covered with integrations tests that check whether:
1. Adding a deletion timestamp to the pod requeues a reconciliation loop for when the grace period elapses.
1. After the grace period elapses, the pod is marked as `Failed`.
1. Replacement pods are scheduled in the place of the failed pod (optional, technically 1+2 is sufficient to prove this).

Existing integration tests should prove that this feature does not impact Kueue during normal operation.

### Graduation Criteria

#### Alpha

> - Positive feedback from the users.

Reviewer: Add feature gate disabled by default

}

type FailureRecoveryPolicy struct {
  // Rules specifies the rules to be enabled for failure recovery.
  // Exactly one rule can be specified. We keep the API flexible to accommodate
  // setting more rules in the future.
  Rules []FailureRecoveryRule `json:"rules"`
}

type FailureRecoveryRule struct {
  // Exactly one of the fields below must be specified.

  // TerminatePod enables and contains configuration for the `TerminatePod` strategy.
  // This strategy recovers stuck pods by forcefully terminating them after a configured
  // grace period elapses.
  // Currently specifying the field is required. We keep the API flexible to allow
  // introducing other rule actions in the future.
  // +optional
  TerminatePod *TerminatePodConfig `json:"terminatePod,omitempty"`
}

type TerminatePodConfig struct {
  // PodLabelSelector specifies the scope of resources covered by `TerminatePod` failure recovery -
  // resources not matching the selector are ignored by the controller.
  PodLabelSelector metav1.LabelSelector `json:"podLabelSelector"`

  // ForcefulTerminationGracePeriod is the duration between when the pod's `deletionGracePeriodSeconds`

>   // elapses and when the pod should be forcefully deleted.

Reviewer: ```suggestion
  // elapses and when the pod should be terminated.
```

  // Represented using metav1.Duration (e.g. "10m", "1h30m").

>   ForcefulTerminationGracePeriod metav1.Duration `json:"forcefulTerminationGracePeriod"`

Author: I missed your last comment - let's go for 3. then.

>   ForcefulTerminationGracePeriod metav1.Duration `json:"forcefulTerminationGracePeriod"`

Author: I added https://github.com/kubernetes-sigs/kueue/pull/7311/commits/8a9628646235e913d44c0b52704d549e14cc32c5 with the change. I'm still looking for any leftovers, but it should be reviewable already.

I'm keeping this discussion open.

>   ForcefulTerminationGracePeriod metav1.Duration `json:"forcefulTerminationGracePeriod"`

Reviewer: I think we can make it an annotation also, annotations are considered more lightweight mechanism. I don't think we need label

- [Design Details](#design-details)
  - [Affected Pods](#affected-pods)
  - [Grace Period](#grace-period)
  - [Implementation Overview](#implementation-overview)
  - [Test Plan](#test-plan)
    - [Unit Tests](#unit-tests)
    - [Integration Tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha](#alpha)
  - [Beta](#beta)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
  - [Do Nothing](#do-nothing)
  - [Control All Pods Managed By Kueue](#control-all-pods-managed-by-kueue)
  - [Introduce a <code>FailureRecoveryPolicy</code> API](#introduce-a-failurerecoverypolicy-api)
  - [Managing The <code>node.kubernetes.io/out-of-service</code> Taint On <code>Node</code>](#managing-the-nodekubernetesioout-of-service-taint-on-node)
<!-- /toc -->

## Summary

This KEP introduces an opt-in mechanism of timeout-based graceful handling of zombie pods,
i.e. pods that are stuck in the `Pending`/`Running` state due to a node-level failure.

## Motivation

Currently, a malfunction of the `kubelet` (or a more general node failure) results in the pods running on that node to become stuck.

> When the `kubelet` fails to send its regular heartbeat to the control plane within `node-monitor-grace-period`, the node is deemed unhealthy and is assigned the `node.kubernetes.io/unreachable` taint. By default, Kubernetes [automatically adds a 5-minute toleration](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions) for pods running in the affected node (this can be explicitly set by a different value by the user/controller).

Reviewer 3: ```suggestion
When the `kubelet` fails to send its regular heartbeat to the control plane within `node-monitor-grace-period`, 
the node is deemed unhealthy and is assigned the `node.kubernetes.io/unreachable` taint (after being marked with `node.kubernetes.io/not-ready` first).
By default, Kubernetes [automatically adds a 5-minute toleration](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions) for pods running in the affected node (this can be explicitly set by a different value by the user/controller).
```

AFAIK, the node taints will be progressing N/A -> `node.kubernetes.io/not-ready` -> `node.kubernetes.io/unreachable`

Author: For reference, it's described [here](https://kubernetes.io/docs/reference/node/node-status/#condition) in the docs.

If the heartbeat is resumed within this toleration period, the control plane removes the taints and the pods continue running.
On the other hand, if it is not resumed, then the pods are marked for termination (i.e. their `deletionTimestamp` is set in `etcd`).

Without a functioning `kubelet`, the pods have no way of progressing beyond that point on their own.
Such pods, colloquially called "zombie pods", remain terminating until the node is healed or they are manually removed.
This is a crucial safety measure, as in that scenario the control plane has no way to confirm that the pods actually stopped
and released all their resources. It is especially relevant for stateful applications and might result in data corruption.

> For example, the stuck pod might still be writing to a `PersistentVolume` when a replacement is started.

Reviewer 3: > It is especially relevant for stateful applications and might result in data corruption.
For example, the stuck pod might still be writing to a `PersistentVolume` when a replacement is started.

This is just a confirmation.
Does this assume that Node is healthy (containerd and runc are healthy), but kubelet is down?

Author: There's also the scenario of network partitions, where everything is running (including the kubelet) but there's no way for the control plane and kubelet to communicate.

I'm wondering whether that's something that should be mentioned explicitly in the KEP as well.

```yaml
kueue.x-k8s.io/safe-to-forcefully-terminate: "true"
```

Only pods containing the new annotation will be affected by the controller, for example:
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  generateName: sample-job-
  namespace: default
  labels:
    kueue.x-k8s.io/queue-name: user-queue
spec:
  parallelism: 3
  completions: 3
  podReplacementPolicy: Failed
  template:
    metadata:
      annotations:
        kueue.x-k8s.io/safe-to-forcefully-terminate: "true"  // <- new annotation
    spec:
      containers:
      - name: dummy-job
        image: registry.k8s.io/e2e-test-images/agnhost:2.53
        command: [ "/bin/sh" ]
        args: [ "-c", "sleep 600" ]
        resources:
          requests:
            cpu: "1"
            memory: "200Mi"
      restartPolicy: Never
```

> The controller is not limited to pods managed by Kueue (i.e. having the `kueue.x-k8s.io/managed` or `kueue.x-k8s.io/podset` label). All pods in the cluster that have the new annotation will be affected.

### Grace Period

A default grace period of **1 minute** will be introduced to get initial feedback about the feature.

> This will help inform the decision on the structure of the API and setting defaults makes sense when graduating to beta.

Reviewer 3: Can you mention that the grace period is counted based on Pod `.metadata.deletionGracePeriodSeconds`?

    * `pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodPending`
1. Are not annotated with the new `kueue.x-k8s.io/safe-to-forcefully-terminate` annotation.
1. Are **not** scheduled on a node tainted with `node.kubernetes.io/unreachable`.
    * This explicitly ignores pods assigned to nodes that still have a running kubelet.
    For example, nodes with the `node.kubernetes.io/not-ready` taint experiencing resource pressure
    that makes pod termination take longer.

For relevant (not ignored) terminating pods, the controller schedules another reconciliation
to happen after the remaining grace period elapses.

```go
func (r *TerminatingPodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
  // ...

  now := r.clock.Now()
  gracefulTerminationPeriod := time.Duration(*pod.DeletionGracePeriodSeconds) * time.Second
  totalGracePeriod := gracefulTerminationPeriod + defaultForcefulTerminationGracePeriod
  if now.Before(pod.DeletionTimestamp.Add(totalGracePeriod)) {
    gracePeriodLeft := pod.DeletionTimestamp.Add(totalGracePeriod).Sub(now)
    return ctrl.Result{RequeueAfter: gracePeriodLeft}, nil
  }

  // ...
}

> ```

Author: Agreed. Adding code did seem awkward, especially when I had to keep it in sync with the implementation. Removed.

In that scheduled reconciliation, unless the node recovered or the pod was deleted,

> the pod will be deemed "zombie" and transitioned into the `PodFailed` phase:

Reviewer: sgtm, also sending an event would be helpful

```go
func (r *ZombiePodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
  // ...

  pod.Status.Phase = corev1.PodFailed
  if err := r.client.Status().Update(ctx, pod); err != nil {
    return ctrl.Result{}, err
  }

  return ctrl.Result{}, nil
}

> ```

Reviewer 3: ```suggestion
```

As described above.

existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit Tests

The proposal will be covered with unit tests for:
1. Configuration parsing.
1. Controller behavior:
    1. Whether it ignores irrelevant pods (not terminating, already failed/succeeded, not annotated).
    1. Whether it correctly schedules a reconciliation for when the grace period elapses.
    1. Whether it updates the pod's phase to `Failed` after the grace period elapses.

#### Integration Tests

The proposal will be covered with integrations tests that check whether:
1. Adding a deletion timestamp to the pod requeues a reconciliation loop for when the grace period elapses.
1. After the grace period elapses, the pod is marked as `Failed`.
1. Replacement pods are scheduled in the place of the failed pod (optional, technically 1+2 is sufficient to prove this).

Existing integration tests should prove that this feature does not impact Kueue during normal operation.

### Graduation Criteria

#### Alpha

- Feature gate disabled by default.
- Positive feedback from the users.

### Beta

- Feature gate enabled by default.

> - Re-evaluate the introduction of the `FailureRecoveryPolicy` API.

Reviewer 3: I agree. We should reevaluate that later.

## Implementation History

2025-09-08: The [issue](https://github.com/kubernetes-sigs/kueue/issues/6757) is raised in Kueue.

2025-09-12: The [issue](https://github.com/kubernetes/kubernetes/issues/134038) is raised in core Kubernetes.

2025-10-17: First draft of the KEP.

## Drawbacks

* The same feature is discussed in core Kubernetes ([kubernetes/issues/134038](https://github.com/kubernetes/kubernetes/issues/134038)), so the underlying issue could potentially be fixed upstream. The timeline of an upstream change is long, but if the feature is deemed not time-critical, it could be fixed at the source instead of in Kueue.
* This feature introduces a potential footgun to Kueue users and could turn out to be to volatile/risky to use in most affected cases.
* It introduces yet another responsibility for Kueue - on top of quota management and scheduling, it will also start performing failure recovery.

> ## Alternatives

Reviewer 3: IIUC, Non Graceful Shutdown (https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#non-graceful-node-shutdown) could be considered as an alternative.
